### ISSUE TITLE
Migration example

### ISSUE THREAD

**Issue by rahuldaga** [dgrijalva/jwt-go#266]
The migration guide  https://github.com/dgrijalva/jwt-go/blob/master/MIGRATION_GUIDE.md
```
type MyCustomClaims struct {
		User string
		*StandardClaims
	}

	if token, err := jwt.ParseWithClaims(tokenString, &MyCustomClaims{}, keyLookupFunc); err == nil {
		claims := token.Claims.(*MyCustomClaims)
		fmt.Printf("Token for user %v expires %v", claims.User, claims.StandardClaims.ExpiresAt)
	}
```

should really be

```
  claims := new (MyCustomClaims)
  if token, err := jwt.ParseWithClaims(tokenString, claims, keyLookupFunc); err == nil {
	fmt.Printf("Token for user %v expires %v", claims.User, claims.StandardClaims.ExpiresAt)
   }
```

**Comment by umpc** [comment]
Hi,

If you look closely, this example uses `err == nil` rather than `err != nil` and the asserted claims represent a value that certainly contains the expected parsed values rather than a struct type value that may or may not only contain zero-values.

**Comment by rahuldaga** [comment]
Apologies if i'm missing something but it looks like claims pointer is the same as the one passed in hence you could hold onto the reference and get rid of the type assertion later. Assuming those claims must exist for a valid token I would  

```
claims := new (MyCustomClaims)
  if token, err := jwt.ParseWithClaims(tokenString, claims, keyLookupFunc); err == nil  {
       if token.valid {
	fmt.Printf("Token for user %v expires %v", claims.User, claims.StandardClaims.ExpiresAt)
       } else {
         // alternate action
        }
   }
```

**Comment by umpc** [comment]
I agree that it is redundant and unusual.

This makes more sense when looking at the [`Token` type](https://godoc.org/github.com/dgrijalva/jwt-go#Token), which contains all of the token data together, raw and parsed.

The abstraction works in the same way that JWTs themselves contains the header (which contains the signing method), claims, and signature. If you have good ideas for improving the type, now is the time, as 4.0.0 is slowly being developed.

**Comment by rahuldaga** [comment]
Yeah that does make sense. If only go would allow concrete dynamic type passed by caller i.e. java generics equivalent. 

Anyway both methods work but FYI I was looking at it idiomatically wherein system pkgs such as json encoders/marshallers/parsers tend to take a data structure pointer to populate rather than return something. 

However, in this case I agree with you as there is no other way really (cant just return claims Decoder back and leave the unmarshalling to the caller as you obviously need to parse standard claims to validate them)
